src/core/ext/transport/chttp2/client/secure/secure_channel_create.c

```diff
@@ -55,8 +56,79 @@ static void client_channel_factory_unref(
 static grpc_subchannel *client_channel_factory_create_subchannel(
     grpc_exec_ctx *exec_ctx, grpc_client_channel_factory *cc_factory,
     const grpc_subchannel_args *args) {
+  grpc_subchannel_args final_sc_args;
+  memcpy(&final_sc_args, args, sizeof(grpc_subchannel_args));
+  const grpc_arg *target_info_arg =
+      grpc_channel_args_find(args->args, GRPC_ARG_GRPCLB_BALANCER_NAMES);
+  grpc_channel_args *new_channel_args = NULL;
+
+  if (target_info_arg != NULL) {
+    // For LB channels, regenerate the security connector. The credentials used
+    // for the new security connection are stripped of call credentials and has
+    // knowledge about the balancer names, which are to be used as the "target"
+    // for the purposes of secure naming.
+    const grpc_channel_targets_info *targets_info =
+        target_info_arg->value.pointer.p;
+
+    // To which address are we connecting?
+    char *target;
+    GPR_ASSERT(
+        grpc_sockaddr_to_string(&target, args->addr, true /* normalize */));
+
+    // Find the balancer name for that target.
+    const char *balancer_name_target =
+        grpc_channel_targets_info_find(targets_info, target);
```

Comment:
Could the target name lookup be done inside of the security handshaker, rather than being done centrally here?  That way, we don't need to do it unless the security handshaker actually needs the information.